common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -214,168 +105,92 @@ public void run() {
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
         //If a smartcard device is connected, proceed with smartcard CBA.
-        synchronized (sDeviceLock) {
-            if (mDevice != null) {
-                handleSmartcardCertAuth(request);
-                return null;
-            }
+        if (mSmartcardCertBasedAuthManager.isDeviceConnected()) {
+            handleSmartcardCertAuth(request);
+            return null;
         }
         //Else, proceed with user certificates stored on device.
         handleOnDeviceCertAuth(request);
```

Comment:
I think it was brought up in my previous PR that maybe regular on-device CBA (which is started in the method `handleOnDeviceCertAuth`) and smartcard CBA should be separate features in their own classes. Are there any thoughts on how this could be done to improve the structure of `ClientCertAuthChallengeHandler`?

Reply:
I would suggest pulling these capabilities into separate interfaces and having the functions/methods which require this behavior accept an instance of that interface, rather than object that can do both

Reply:
(If a given class can do both, it should implement both of those interfaces)

Reply:
If I understand correctly, this would mean creating an interface for on-device CBA and a separate one for smartcard CBA, and then feeding in implementations of these interfaces to `ClientCertAuthChallengeHandler` as parameters? This would leave `ClientCertAuthChallengeHandler` without the actual implementations of the features, so it can just call the interfaces to handle CBA as necessary.

Reply:
I might have given clumsy feedback in my initial review.... trying again.

My.... far at a distance.... view would be that "CBA is CBA".... sometimes it happens on device, and sometimes it happens via smartcard

You might want a generic "Can-perform-CBA" interface and have different implementations of that interface which know how to do it on-device and via smartcard. Then.... pass the appropriate instance at runtime based on which you're doing.

If they accept/need different params, you can hide this detail in a parameters object that can work across either class.

So to summarize with a pseudocode example....

```java
public interface ICertBasedAuthHandler {
    void performCba(ClientCertRequest request);
}

public class OnDeviceCbaHandler implements ICertBasedAuthHandler {
    @Override
    void performCba(ClientCertRequest request) {
         // provide implementation
    }
}

public class SmartCardCbaHandler implements ICertBasedAuthHandler {
    @Override
    void performCba(ClientCertRequest request) {
         // provide implementation
    }
}

// Then.... hide the if/else behind object construction... A factory can help here....
public static ICertBasedAuthHandler createCertBasedAuthHandler(boolean isConnected) {
    return isConnected ? new SmartCardCbaHandler() : new OnDeviceCbaHandler();
}

// Put it all together and your method starts to look like this....
@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
@Override
public Void processChallenge(@NonNull final ClientCertRequest request) {
    final ICertBasedAuthHandler cbaHandler = createCertBasedAuthHandler(mSmartcardCertBasedAuthManager);
    cbaHandler.performCba(request);
    return null;
}
```